### ISSUE TITLE
dependabot update command fails for private repo

### ISSUE THREAD

**Issue by OrRosenblatt** [dependabot/cli#123]
I've installed the dependabot CLI and run `dependabot update -f job.yml` based on the [provided example](https://github.com/dependabot/cli/blob/main/testdata/private-npm.yaml).  
However, the job fails with `unknown_error`.  

<details>
<summary>Job Spec</summary>

The `job.yml`, based on the [provided example](https://github.com/dependabot/cli/blob/main/testdata/private-npm.yaml):
```
job:
  package-manager: npm_and_yarn
  allowed-updates:
    - update-type: all
  source:
    provider: github
    repo: <my-org>/<my-private-repo>
    directory: /

credentials:
  - type: npm_registry
    registry: https://npm.pkg.github.com
    token: $LOCAL_GITHUB_ACCESS_TOKEN
```

</details>

<details>
<summary>Result Output</summary>

This is the output of `LOCAL_GITHUB_ACCESS_TOKEN=<my_local_gh_access_token> dependabot update -f job.yml` (**_org_** and **_repo_** have been masked):

```log
    cli | 2023/05/23 04:48:12 Inserting $LOCAL_GITHUB_ACCESS_TOKEN into credentials
    cli | 2023/05/23 04:48:12 Adding missing credentials-metadata into job definition
    cli | 2023/05/23 04:48:12 using image ghcr.io/github/dependabot-update-job-proxy/dependabot-update-job-proxy:latest at sha256:3810d68f41de0398ed2280c5e27f7ae4fd6a9495cddf18535e0058a1ab628201
    cli | 2023/05/23 04:48:12 using image ghcr.io/dependabot/dependabot-updater-npm at sha256:a660ae0c85f126e2ec5cf127b64ebc4ae0cbfd6f525555569eb91cfecaac9b25
  proxy | 2023/05/23 04:48:13 proxy starting, commit: 1a083db3c26dfa9c07829145a6949d6e525c7ab1
  proxy | 2023/05/23 04:48:13 initializing metrics client: No address passed and autodetection from environment failed
  proxy | 2023/05/23 04:48:13 Listening (:1080)
updater | Updating certificates in /etc/ssl/certs...
updater | rehash: warning: skipping ca-certificates.crt,it does not contain exactly one certificate or CRL
updater | 1 added, 0 removed; done.
updater | Running hooks in /etc/ca-certificates/update.d...
updater | done.
updater | 2023/05/23 04:48:15 INFO Raven 3.1.2 configured not to capture errors: DSN not set
updater | 2023/05/23 04:48:16 INFO Starting job processing
  proxy | 2023/05/23 04:48:16 [002] GET https://github.com:443/<my-org>/<my-private-repo>/info/refs?service=git-upload-pack
  proxy | 2023/05/23 04:48:16 [002] * authenticating git server request (host: github.com)
  proxy | 2023/05/23 04:48:16 [002] 200 https://github.com:443/<my-org>/<my-private-repo>/info/refs?service=git-upload-pack
  proxy | 2023/05/23 04:48:16 [004] POST https://github.com:443/<my-org>/<my-private-repo>/git-upload-pack
  proxy | 2023/05/23 04:48:16 [004] * authenticating git server request (host: github.com)
  proxy | 2023/05/23 04:48:16 [004] 200 https://github.com:443/<my-org>/<my-private-repo>/git-upload-pack
  proxy | 2023/05/23 04:48:16 [006] POST https://github.com:443/<my-org>/<my-private-repo>/git-upload-pack
  proxy | 2023/05/23 04:48:16 [006] * authenticating git server request (host: github.com)
  proxy | 2023/05/23 04:48:17 [006] 200 https://github.com:443/<my-org>/<my-private-repo>/git-upload-pack
  proxy | 2023/05/23 04:48:17 [007] POST http://host.docker.internal:53194/update_jobs/cli/record_package_manager_version
{"data":{"ecosystem":"npm","package-managers":{"yarn":"3.5.1"}},"type":"record_package_manager_version"}
  proxy | 2023/05/23 04:48:17 [007] 200 http://host.docker.internal:53194/update_jobs/cli/record_package_manager_version
updater | 2023/05/23 04:48:17 INFO Finished job processing
updater | 2023/05/23 04:48:18 INFO Raven 3.1.2 configured not to capture errors: DSN not set
updater | 2023/05/23 04:48:18 INFO Starting job processing
updater | 2023/05/23 04:48:18 ERROR undefined method `length' for nil:NilClass
updater | 
updater |           sort_by { |cred| cred["registry"].length }.
updater |                                            ^^^^^^^
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:297:in `block in url_for_relevant_cred'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:297:in `each'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:297:in `sort_by'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:297:in `url_for_relevant_cred'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:283:in `registry_source_for'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:236:in `source_for'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:119:in `build_dependency'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:71:in `block (3 levels) in manifest_dependencies'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:67:in `each'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:67:in `block (2 levels) in manifest_dependencies'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:65:in `each'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:65:in `block in manifest_dependencies'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:59:in `each'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:59:in `manifest_dependencies'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:37:in `parse'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/dependabot-updater/lib/dependabot/dependency_snapshot.rb:97:in `parse_files!'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/dependabot-updater/lib/dependabot/dependency_snapshot.rb:91:in `initialize'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/dependabot-updater/lib/dependabot/dependency_snapshot.rb:21:in `new'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/dependabot-updater/lib/dependabot/dependency_snapshot.rb:21:in `create_from_job_definition'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/dependabot-updater/lib/dependabot/update_files_command.rb:16:in `perform_job'
updater | 2023/05/23 04:48:18 ERROR /home/dependabot/dependabot-updater/lib/dependabot/base_command.rb:52:in `run'
updater | 2023/05/23 04:48:18 ERROR bin/update_files.rb:23:in `<main>'
  proxy | 2023/05/23 04:48:18 [008] POST http://host.docker.internal:53194/update_jobs/cli/record_update_job_error
{"data":{"error-type":"unknown_error","error-details":null},"type":"record_update_job_error"}
  proxy | 2023/05/23 04:48:18 [008] 200 http://host.docker.internal:53194/update_jobs/cli/record_update_job_error
  proxy | 2023/05/23 04:48:18 [009] PATCH http://host.docker.internal:53194/update_jobs/cli/mark_as_processed
{"data":{"base-commit-sha":"ce90be394d6c90e19cdc5237e60ff738152c14cd"},"type":"mark_as_processed"}
updater | 2023/05/23 04:48:18 INFO Finished job processing
  proxy | 2023/05/23 04:48:18 [009] 200 http://host.docker.internal:53194/update_jobs/cli/mark_as_processed
updater | 2023/05/23 04:48:18 INFO Results:
updater | Dependabot encountered '1' error(s) during execution, please check the logs for more details.
updater | +---------------+
updater | |    Errors     |
updater | +---------------+
updater | | unknown_error |
updater | +---------------+
  proxy | 2023/05/23 04:48:19 0/3 calls cached (0%)
```

</details>

<details>
<summary>Docker data</summary>

`docker version`:
```
Client:
 Cloud integration: v1.0.31
 Version:           20.10.23
 API version:       1.41
 Go version:        go1.18.10
 Git commit:        7155243
 Built:             Thu Jan 19 17:35:19 2023
 OS/Arch:           darwin/amd64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.17.0 (99724)
 Engine:
  Version:          20.10.23
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.10
  Git commit:       6051f14
  Built:            Thu Jan 19 17:32:04 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

</details>

Can you please help me understand the issue?

**Comment by jakecoffman** [comment]
Thanks for reporting this!

I have a fix in #125. 

In the meantime you can work around this by setting a `credentials-metadata` section in the job definition:

```yaml
job:
  package-manager: npm_and_yarn
  allowed-updates:
    - update-type: all
  source:
    provider: github
    repo: <my-org>/<my-private-repo>
    directory: /
  credentials-metadata:
    - type: npm_registry
      registry: https://npm.pkg.github.com

credentials:
  - type: npm_registry
    registry: https://npm.pkg.github.com
    token: $LOCAL_GITHUB_ACCESS_TOKEN
```

**Comment by jakecoffman** [comment]
Released in https://github.com/dependabot/cli/releases/tag/v1.25.1
